[nnc]  Added a transformation to move users of aten::cat to its inputs in the NNC kernel subgraph.

navahgar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #59581 [nnc] Added micro-benchmark to show perf improvement with cat subgraph optimization
* **#59580 [nnc]  Added a transformation to move users of aten::cat to its inputs in the NNC kernel subgraph.**
* #60162 [jit] Exported a method to get the supported list of elementwise ops
* #59579 [jit] Made a list for element-wise ops.

Differential Revision: [D28955318](https://our.internmc.facebook.com/intern/diff/D28955318)

navahgar committed:
[jit] Added a tranformation to move consumers of aten::cat to its inputs, in the fused subgraphs

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28955318).

eellison commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -1197,6 +1197,113 @@ class TensorExprFuser {
+    for (auto inp : cat_inputs) {
+      auto new_cat_input = subgraph->createClone(
+          user, [&](Value* k) { return (k == cat->output()) ? inp : k; });
+      new_cat_input->output()->setType(inp->type());
```
is it possible we want the dtype of the output of cat here ? In case we have something like `sigmoid` which has type promotion ?

eellison commented on torch/csrc/jit/passes/tensorexpr_fuser.cpp:
```diff
@@ -1197,6 +1197,113 @@ class TensorExprFuser {
+    user->output()->replaceAllUsesWith(new_cat->output());
+    user->destroy();
+
+    TORCH_INTERNAL_ASSERT(!cat->output()->hasUses());
```
just fwiw `cat->destroy()` will also check that the output has no uses, but ya maybe it's nice to be explicit here with the assertion before

eellison reviewed (commented):
Nice, looks good to me!! should be about ready to land just had a question or two

navahgar committed:
Update on "[jit] Added a tranformation to move consumers of aten::cat to its inputs, in the fused subgraphs"


Differential Revision: [D28955318](https://our.internmc.facebook.com/intern/diff/D28955318)

[ghstack-poisoned]

navahgar reviewed (commented):
Thanks for the review @eellison. Addressed your comments.

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28955318).

bertmaher commented on test/cpp/tensorexpr/test_graph_opt.cpp:
```diff
@@ -0,0 +1,322 @@
+      ->check_not("aten::log")
+      ->run(*kernel.graph());
+
+  auto x = at::rand({10}, TensorOptions(kCPU).dtype(at::kFloat));
```
I think you can just do `at::rand({10}, at::kFloat)`

navahgar replied:
Done

bertmaher commented on torch/csrc/jit/tensorexpr/graph_opt.cpp:
```diff
@@ -0,0 +1,170 @@
+  for (auto inp : cat_inputs) {
+    auto new_cat_input = subgraph->createClone(
+        user, [&](Value* k) { return (k == cat->output()) ? inp : k; });
+    // The type of new input should be same as the type of the user. This is
```
By `type` you mean specifically `scalar type` here, right, since obviously the size of the input and the size of the user will be different?

navahgar replied:
This comment was left over from some earlier changes I was making. Removed now.

bertmaher commented on torch/csrc/jit/tensorexpr/graph_opt.cpp:
```diff
@@ -0,0 +1,170 @@
+    // promotion.
+    // new_cat_input->output()->setType(user->output()->type());
+
+    // Since we are cloning user, its result should be the same scalar type
```
Is this comment saying basically the same thing as the one above it?  Can you merge them?

navahgar replied:
Removed the previous comment.

navahgar committed:
Update on "[nnc]  Added a transformation to move users of aten::cat to its inputs in the NNC kernel subgraph."


Differential Revision: [D28955318](https://our.internmc.facebook.com/intern/diff/D28955318)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28955318).

bertmaher commented on torch/csrc/jit/tensorexpr/graph_opt.cpp:
```diff
@@ -0,0 +1,164 @@
+  new_cat->insertBefore(cat);
+
+  user->output()->replaceAllUsesWith(new_cat->output());
+  user->destroy();
+
+  TORCH_INTERNAL_ASSERT(!cat->output()->hasUses());
+  cat->destroy();
+
+  if (!cat_list->output()->hasUses()) {
+    cat_list->destroy();
+  }
```
Maybe a question for @eellison  too: do we actually have to explicitly destroy the cat/listconstruct, or will those get cleaned up by dce or something?

navahgar commented on test/cpp/tensorexpr/test_graph_opt.cpp:
```diff
@@ -0,0 +1,322 @@
+
+// NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables)
+TEST(GraphOpt, OptimizeCat) {
+  bool old_flag = getCatWoConditionals();
```
Sure, done.

navahgar committed:
Update on "[nnc]  Added a transformation to move users of aten::cat to its inputs in the NNC kernel subgraph."


Differential Revision: [D28955318](https://our.internmc.facebook.com/intern/diff/D28955318)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28955318).

navahgar committed:
Update on "[nnc]  Added a transformation to move users of aten::cat to its inputs in the NNC kernel subgraph."


Differential Revision: [D28955318](https://our.internmc.facebook.com/intern/diff/D28955318)

[ghstack-poisoned]

navahgar commented:
@navahgar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28955318).
